Sending response although request has been cancelled

## Description

Even if an XHR has been cancelled PollyJS will still try to send a response. 
This will lead to an error. 

### Shareable Source

```js
    const unsubscribe = new Subject();
    this.httpClient.get('/rest/my-api').pipe(takeUntil(unsubscribe)).subscribe((result) => console.log(result));
    unsubscribe.next();
    unsubscribe.complete();
```

PollsJS will still call
```js
xhr.respond(response.statusCode, response.headers, response.body);
```
on the already cancelled request resulting in an error.

I think the issue can be fixed in the XHRAdapter:
```js
if (!xhr.aborted) {
xhr.respond(response.statusCode, response.headers, response.body);
}
```

### Error Message & Stack Trace

```
pollyjs-core.js:7258 mock
pollyjs-core.js:7284 Errored ➞ GET /rest/my-api
pollyjs-core.js:7285 Error: INVALID_STATE_ERR - 0
    at verifyRequestOpened (pollyjs-adapter-xhr.js:6788)
    at FakeXMLHttpRequest.setStatus (pollyjs-adapter-xhr.js:7028)
    at FakeXMLHttpRequest.respond (pollyjs-adapter-xhr.js:7193)
    at XHRAdapter.respondToRequest (pollyjs-adapter-xhr.js:11077)
    at XHRAdapter._callee14$ (pollyjs-adapter-xhr.js:10865)
    at tryCatch (pollyjs-adapter-xhr.js:68)
    at Generator.invoke [as _invoke] (pollyjs-adapter-xhr.js:294)
    at Generator.prototype.<computed> [as next] (pollyjs-adapter-xhr.js:120)
    at asyncGeneratorStep (pollyjs-adapter-xhr.js:1961)
    at _next (pollyjs-adapter-xhr.js:1983)
```

### Config

```js
      Polly.register(XHRAdapter);

      const polly = new Polly('mock', {
        adapters: ['xhr']
      });

      const { server } = polly;

      server.get('rest/my-api').intercept((req, res) => { res.send({}); });
```

### Dependencies

```json
{
    "@pollyjs/adapter-xhr": "~4.0.2",
    "@pollyjs/core": "~4.0.2",
}
```

## Relevant Links

https://github.com/nbfr/pollyjs-dream-app
Branch: test/cancelled-request

Run "npm run start-mock" and open the app on "http://localhost:4200" to see the error in the browser console.

## Environment

Node.JS v10.16.3
win32 10.0.1832
npm 6.9.0

--- a/docs/server/events-and-middleware.md
+++ b/docs/server/events-and-middleware.md
@@ -150,6 +150,24 @@ server.any().on('error', (req, error) => {
 });
 ```
 
+### abort
+
+Fires when a request is aborted.
+
+| Param | Type                      | Description          |
+| ----- | ------------------------- | -------------------- |
+| req   | [Request](server/request) | The request instance |
+| event | [Event](server/event)     | The event instance   |
+
+**Example**
+
+```js
+server.any().on('abort', req => {
+  console.error('Request aborted.');
+  process.exit(1);
+});
+```
+
 ## Middleware
 
 Middleware can be added via the `.any()` method.

--- a/packages/@pollyjs/adapter-fetch/src/index.js
+++ b/packages/@pollyjs/adapter-fetch/src/index.js
@@ -134,6 +134,18 @@ export default class FetchAdapter extends Adapter {
     this.NativeRequest = null;
   }
 
+  onRequest(pollyRequest) {
+    const { options } = pollyRequest.requestArguments;
+
+    if (options.signal) {
+      if (options.signal.aborted) {
+        pollyRequest.abort();
+      } else {
+        options.signal.addEventListener('abort', () => pollyRequest.abort());
+      }
+    }
+  }
+
   async passthroughRequest(pollyRequest) {
     const { context } = this.options;
     const { options } = pollyRequest.requestArguments;
@@ -160,6 +172,17 @@ export default class FetchAdapter extends Adapter {
       context: { Response }
     } = this.options;
     const { respond } = pollyRequest.requestArguments;
+
+    if (pollyRequest.aborted) {
+      respond({
+        error: new DOMException(
+          'The user aborted a request.',
+          DOMException.ABORT_ERR
+        )
+      });
+
+      return;
+    }
 
     if (error) {
       respond({ error });

--- a/packages/@pollyjs/adapter-fetch/src/index.js
+++ b/packages/@pollyjs/adapter-fetch/src/index.js
@@ -5,6 +5,7 @@ import serializeHeaders from './utils/serializer-headers';
 
 const { defineProperty } = Object;
 const IS_STUBBED = Symbol();
+const ABORT_HANDLER = Symbol();
 const REQUEST_ARGUMENTS = Symbol();
 
 export default class FetchAdapter extends Adapter {
@@ -135,13 +136,16 @@ export default class FetchAdapter extends Adapter {
   }
 
   onRequest(pollyRequest) {
-    const { options } = pollyRequest.requestArguments;
+    const {
+      options: { signal }
+    } = pollyRequest.requestArguments;
 
-    if (options.signal) {
-      if (options.signal.aborted) {
+    if (signal) {
+      if (signal.aborted) {
         pollyRequest.abort();
       } else {
-        options.signal.addEventListener('abort', () => pollyRequest.abort());
+        pollyRequest[ABORT_HANDLER] = () => pollyRequest.abort();
+        signal.addEventListener('abort', pollyRequest[ABORT_HANDLER]);
       }
     }
   }
@@ -171,14 +175,18 @@ export default class FetchAdapter extends Adapter {
     const {
       context: { Response }
     } = this.options;
-    const { respond } = pollyRequest.requestArguments;
+    const {
+      respond,
+      options: { signal }
+    } = pollyRequest.requestArguments;
+
+    if (signal && pollyRequest[ABORT_HANDLER]) {
+      signal.removeEventListener('abort', pollyRequest[ABORT_HANDLER]);
+    }
 
     if (pollyRequest.aborted) {
       respond({
-        error: new DOMException(
-          'The user aborted a request.',
-          DOMException.ABORT_ERR
-        )
+        error: new DOMException('The user aborted a request.', 'AbortError')
       });
 
       return;

--- a/packages/@pollyjs/adapter-fetch/tests/integration/adapter-test.js
+++ b/packages/@pollyjs/adapter-fetch/tests/integration/adapter-test.js
@@ -53,6 +53,33 @@ describe('Integration | Fetch Adapter', function() {
     expect(headers).to.deep.equal({ 'content-type': 'application/json' });
   });
 
+  it('should handle aborting a request', async function() {
+    const { server } = this.polly;
+    const controller = new AbortController();
+    let request;
+    let error;
+
+    server
+      .any(this.recordUrl())
+      .on('abort', req => (request = req))
+      .intercept(async (_, res) => {
+        await server.timeout(50);
+        res.sendStatus(200);
+      });
+
+    try {
+      const promise = this.fetchRecord({ signal: controller.signal });
+
+      setTimeout(() => controller.abort(), 25);
+      await promise;
+    } catch (e) {
+      error = e;
+    }
+
+    expect(request.aborted).to.equal(true);
+    expect(error.message).to.contain('The user aborted a request.');
+  });
+
   describe('Request', function() {
     it('should support Request objects', async function() {
       const { server } = this.polly;

--- a/packages/@pollyjs/adapter-fetch/tests/integration/adapter-test.js
+++ b/packages/@pollyjs/adapter-fetch/tests/integration/adapter-test.js
@@ -56,27 +56,50 @@ describe('Integration | Fetch Adapter', function() {
   it('should handle aborting a request', async function() {
     const { server } = this.polly;
     const controller = new AbortController();
-    let request;
+    let abortEventCalled = false;
     let error;
 
     server
       .any(this.recordUrl())
-      .on('abort', req => (request = req))
-      .intercept(async (_, res) => {
-        await server.timeout(50);
+      .on('request', () => controller.abort())
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
+        res.sendStatus(200);
+      });
+
+    try {
+      await this.fetchRecord({ signal: controller.signal });
+    } catch (e) {
+      error = e;
+    }
+
+    expect(abortEventCalled).to.equal(true);
+    expect(error.message).to.contain('The user aborted a request.');
+  });
+
+  it('should handle immediately aborting a request', async function() {
+    const { server } = this.polly;
+    const controller = new AbortController();
+    let abortEventCalled = false;
+    let error;
+
+    server
+      .any(this.recordUrl())
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
         res.sendStatus(200);
       });
 
     try {
       const promise = this.fetchRecord({ signal: controller.signal });
 
-      setTimeout(() => controller.abort(), 25);
+      controller.abort();
       await promise;
     } catch (e) {
       error = e;
     }
 
-    expect(request.aborted).to.equal(true);
+    expect(abortEventCalled).to.equal(true);
     expect(error.message).to.contain('The user aborted a request.');
   });
 

--- a/packages/@pollyjs/adapter-node-http/src/index.js
+++ b/packages/@pollyjs/adapter-node-http/src/index.js
@@ -151,6 +151,16 @@ export default class HttpAdapter extends Adapter {
     });
   }
 
+  onRequest(pollyRequest) {
+    const { req } = pollyRequest.requestArguments;
+
+    if (req.aborted) {
+      pollyRequest.abort();
+    } else {
+      req.once('abort', () => pollyRequest.abort());
+    }
+  }
+
   async passthroughRequest(pollyRequest) {
     const { parsedArguments } = pollyRequest.requestArguments;
     const { method, headers, body } = pollyRequest;
@@ -195,6 +205,15 @@ export default class HttpAdapter extends Adapter {
 
   async respondToRequest(pollyRequest, error) {
     const { req, respond } = pollyRequest.requestArguments;
+    const { statusCode, body, headers } = pollyRequest.response;
+
+    if (pollyRequest.aborted) {
+      // Even if the request has been aborted, we need to respond to the nock
+      // request in order to resolve its awaiting promise.
+      respond(null, [statusCode, undefined, headers]);
+
+      return;
+    }
 
     if (error) {
       // If an error was received then forward it over to nock so it can
@@ -204,7 +223,6 @@ export default class HttpAdapter extends Adapter {
       return;
     }
 
-    const { statusCode, body, headers } = pollyRequest.response;
     const chunks = this.getChunksFromBody(body, headers);
     const stream = new ReadableStream();
 
@@ -219,13 +237,9 @@ export default class HttpAdapter extends Adapter {
     // that the deferred promise used by `polly.flush()` doesn't resolve before
     // the response was actually received.
     const requestFinishedPromise = new Promise(resolve => {
-      if (req.aborted) {
-        resolve();
-      } else {
-        req.once('response', resolve);
-        req.once('abort', resolve);
-        req.once('error', resolve);
-      }
+      req.once('response', resolve);
+      req.once('abort', resolve);
+      req.once('error', resolve);
     });
 
     respond(null, [statusCode, stream, headers]);

--- a/packages/@pollyjs/adapter-node-http/src/index.js
+++ b/packages/@pollyjs/adapter-node-http/src/index.js
@@ -17,6 +17,7 @@ import mergeChunks from './utils/merge-chunks';
 import urlToOptions from './utils/url-to-options';
 
 const IS_STUBBED = Symbol();
+const ABORT_HANDLER = Symbol();
 const REQUEST_ARGUMENTS = new WeakMap();
 
 // nock begins to intercept network requests on import which is not the
@@ -157,7 +158,8 @@ export default class HttpAdapter extends Adapter {
     if (req.aborted) {
       pollyRequest.abort();
     } else {
-      req.once('abort', () => pollyRequest.abort());
+      pollyRequest[ABORT_HANDLER] = () => pollyRequest.abort();
+      req.once('abort', pollyRequest[ABORT_HANDLER]);
     }
   }
 
@@ -207,10 +209,14 @@ export default class HttpAdapter extends Adapter {
     const { req, respond } = pollyRequest.requestArguments;
     const { statusCode, body, headers } = pollyRequest.response;
 
+    if (pollyRequest[ABORT_HANDLER]) {
+      req.off('abort', pollyRequest[ABORT_HANDLER]);
+    }
+
     if (pollyRequest.aborted) {
       // Even if the request has been aborted, we need to respond to the nock
       // request in order to resolve its awaiting promise.
-      respond(null, [statusCode, undefined, headers]);
+      respond(null, [0, undefined, {}]);
 
       return;
     }
@@ -237,9 +243,13 @@ export default class HttpAdapter extends Adapter {
     // that the deferred promise used by `polly.flush()` doesn't resolve before
     // the response was actually received.
     const requestFinishedPromise = new Promise(resolve => {
-      req.once('response', resolve);
-      req.once('abort', resolve);
-      req.once('error', resolve);
+      if (req.aborted) {
+        resolve();
+      } else {
+        req.once('response', resolve);
+        req.once('abort', resolve);
+        req.once('error', resolve);
+      }
     });
 
     respond(null, [statusCode, stream, headers]);

--- a/packages/@pollyjs/adapter-node-http/tests/integration/adapter-test.js
+++ b/packages/@pollyjs/adapter-node-http/tests/integration/adapter-test.js
@@ -146,4 +146,55 @@ function commonTests(transport) {
 
     expect(nativeHash).to.equal(recordedHash);
   });
+
+  it('should handle aborting a request', async function() {
+    const { server } = this.polly;
+    const url = `${protocol}//example.com`;
+    const req = transport.request(url);
+    let abortEventCalled = false;
+
+    server
+      .any(url)
+      .on('request', () => req.abort())
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
+        res.sendStatus(200);
+      });
+
+    try {
+      await getResponseFromRequest(req);
+    } catch (e) {
+      // no-op
+    }
+
+    await this.polly.flush();
+    expect(abortEventCalled).to.equal(true);
+  });
+
+  it('should handle immediately aborting a request', async function() {
+    const { server } = this.polly;
+    const url = `${protocol}//example.com`;
+    const req = transport.request(url);
+    let abortEventCalled = false;
+
+    server
+      .any(url)
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
+        res.sendStatus(200);
+      });
+
+    const promise = getResponseFromRequest(req);
+
+    req.abort();
+
+    try {
+      await promise;
+    } catch (e) {
+      // no-op
+    }
+
+    await this.polly.flush();
+    expect(abortEventCalled).to.equal(true);
+  });
 }

--- a/packages/@pollyjs/adapter-node-http/tests/utils/get-response-from-request.js
+++ b/packages/@pollyjs/adapter-node-http/tests/utils/get-response-from-request.js
@@ -2,6 +2,7 @@ export default function getResponseFromRequest(req, data) {
   return new Promise((resolve, reject) => {
     req.once('response', resolve);
     req.once('error', reject);
+    req.once('abort', reject);
 
     req.end(data);
   });

--- a/packages/@pollyjs/adapter-xhr/src/index.js
+++ b/packages/@pollyjs/adapter-xhr/src/index.js
@@ -5,6 +5,7 @@ import resolveXhr from './utils/resolve-xhr';
 import serializeResponseHeaders from './utils/serialize-response-headers';
 
 const SEND = Symbol();
+const ABORT_HANDLER = Symbol();
 const stubbedXhrs = new WeakSet();
 
 export default class XHRAdapter extends Adapter {
@@ -67,12 +68,17 @@ export default class XHRAdapter extends Adapter {
     if (xhr.aborted) {
       pollyRequest.abort();
     } else {
-      xhr.addEventListener('abort', () => pollyRequest.abort());
+      pollyRequest[ABORT_HANDLER] = () => pollyRequest.abort();
+      xhr.addEventListener('abort', pollyRequest[ABORT_HANDLER]);
     }
   }
 
   respondToRequest(pollyRequest, error) {
     const { xhr } = pollyRequest.requestArguments;
+
+    if (pollyRequest[ABORT_HANDLER]) {
+      xhr.removeEventListener('abort', pollyRequest[ABORT_HANDLER]);
+    }
 
     if (pollyRequest.aborted) {
       return;

--- a/packages/@pollyjs/adapter-xhr/src/index.js
+++ b/packages/@pollyjs/adapter-xhr/src/index.js
@@ -61,10 +61,22 @@ export default class XHRAdapter extends Adapter {
     this.xhr.restore();
   }
 
+  onRequest(pollyRequest) {
+    const { xhr } = pollyRequest.requestArguments;
+
+    if (xhr.aborted) {
+      pollyRequest.abort();
+    } else {
+      xhr.addEventListener('abort', () => pollyRequest.abort());
+    }
+  }
+
   respondToRequest(pollyRequest, error) {
     const { xhr } = pollyRequest.requestArguments;
 
-    if (error) {
+    if (pollyRequest.aborted) {
+      return;
+    } else if (error) {
       // If an error was received then call the `error` method on the fake XHR
       // request provided by nise which will simulate a network error on the request.
       // The onerror handler will be called and the status will be 0.
@@ -78,7 +90,7 @@ export default class XHRAdapter extends Adapter {
   }
 
   async passthroughRequest(pollyRequest) {
-    const fakeXhr = pollyRequest.requestArguments.xhr;
+    const { xhr: fakeXhr } = pollyRequest.requestArguments;
     const xhr = new this.NativeXMLHttpRequest();
 
     xhr.open(

--- a/packages/@pollyjs/adapter-xhr/tests/integration/adapter-test.js
+++ b/packages/@pollyjs/adapter-xhr/tests/integration/adapter-test.js
@@ -17,16 +17,52 @@ describe('Integration | XHR Adapter', function() {
     persister: InMemoryPersister
   });
 
-  setupFetchRecord({
-    fetch() {
-      return xhrRequest(...arguments);
-    }
-  });
+  setupFetchRecord({ fetch: xhrRequest });
   setupPolly.afterEach();
 
   adapterTests();
   adapterBrowserTests();
   adapterIdentifierTests();
+
+  it('should handle aborting a request', async function() {
+    const { server } = this.polly;
+    const xhr = new XMLHttpRequest();
+    let abortEventCalled;
+
+    server
+      .any(this.recordUrl())
+      .on('request', () => xhr.abort())
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
+        res.sendStatus(200);
+      });
+
+    await this.fetchRecord({ xhr });
+    await this.polly.flush();
+
+    expect(abortEventCalled).to.equal(true);
+  });
+
+  it('should handle immediately aborting a request', async function() {
+    const { server } = this.polly;
+    const xhr = new XMLHttpRequest();
+    let abortEventCalled;
+
+    server
+      .any(this.recordUrl())
+      .on('abort', () => (abortEventCalled = true))
+      .intercept((_, res) => {
+        res.sendStatus(200);
+      });
+
+    const promise = this.fetchRecord({ xhr });
+
+    xhr.abort();
+    await promise;
+    await this.polly.flush();
+
+    expect(abortEventCalled).to.equal(true);
+  });
 });
 
 describe('Integration | XHR Adapter | Init', function() {

--- a/packages/@pollyjs/adapter-xhr/tests/utils/xhr-request.js
+++ b/packages/@pollyjs/adapter-xhr/tests/utils/xhr-request.js
@@ -2,7 +2,7 @@ import serializeResponseHeaders from '../../src/utils/serialize-response-headers
 
 export default function request(url, obj = {}) {
   return new Promise(resolve => {
-    const xhr = new XMLHttpRequest();
+    const xhr = obj.xhr || new XMLHttpRequest();
 
     xhr.open(obj.method || 'GET', url);
 

--- a/packages/@pollyjs/adapter/src/index.js
+++ b/packages/@pollyjs/adapter/src/index.js
@@ -71,15 +71,20 @@ export default class Adapter {
     try {
       pollyRequest.on('identify', (...args) => this.onIdentifyRequest(...args));
 
-      await pollyRequest.setup();
       await this.onRequest(pollyRequest);
+      await pollyRequest.init();
       await this[REQUEST_HANDLER](pollyRequest);
-      await this.onRequestFinished(pollyRequest);
 
-      return pollyRequest;
+      if (pollyRequest.aborted) {
+        throw new PollyError('Request aborted.');
+      }
+
+      await this.onRequestFinished(pollyRequest);
     } catch (error) {
       await this.onRequestFailed(pollyRequest, error);
     }
+
+    return pollyRequest;
   }
 
   async [REQUEST_HANDLER](pollyRequest) {
@@ -228,7 +233,7 @@ export default class Adapter {
 
   /**
    * @param {PollyRequest} pollyRequest
-   * @returns {{ statusCode: number, headers: Object, body: string }}
+   * @returns {Object({ statusCode: number, headers: Object, body: string })}
    */
   async passthroughRequest(/* pollyRequest */) {
     this.assert('Must implement the `passthroughRequest` hook.');
@@ -241,15 +246,19 @@ export default class Adapter {
    * Calling `pollyjs.flush()` will await this method.
    *
    * @param {PollyRequest} pollyRequest
+   * @param {Error} [error]
    */
-  async respondToRequest(/* pollyRequest */) {}
+  async respondToRequest(/* pollyRequest, error */) {}
 
   /**
    * @param {PollyRequest} pollyRequest
    */
   async onRecord(pollyRequest) {
     await this.onPassthrough(pollyRequest);
-    await this.persister.recordRequest(pollyRequest);
+
+    if (!pollyRequest.aborted) {
+      await this.persister.recordRequest(pollyRequest);
+    }
   }
 
   /**
@@ -300,19 +309,25 @@ export default class Adapter {
    */
   async onRequestFinished(pollyRequest) {
     await this.respondToRequest(pollyRequest);
-
     pollyRequest.promise.resolve();
   }
 
   /**
    * @param {PollyRequest} pollyRequest
-   * @param {Error} error
+   * @param {Error} [error]
    */
   async onRequestFailed(pollyRequest, error) {
+    const { aborted } = pollyRequest;
+
     error = error || new PollyError('Request failed due to an unknown error.');
 
     try {
-      await pollyRequest._emit('error', error);
+      if (aborted) {
+        await pollyRequest._emit('abort');
+      } else {
+        await pollyRequest._emit('error', error);
+      }
+
       await this.respondToRequest(pollyRequest, error);
     } catch (e) {
       // Rethrow any error not handled by `respondToRequest`.

--- a/packages/@pollyjs/adapter/src/index.js
+++ b/packages/@pollyjs/adapter/src/index.js
@@ -91,6 +91,10 @@ export default class Adapter {
     const { mode } = this.polly;
     const { _interceptor: interceptor } = pollyRequest;
 
+    if (pollyRequest.aborted) {
+      return;
+    }
+
     if (pollyRequest.shouldIntercept) {
       await this.intercept(pollyRequest, interceptor);
 

--- a/packages/@pollyjs/core/src/-private/request.js
+++ b/packages/@pollyjs/core/src/-private/request.js
@@ -38,6 +38,7 @@ export default class PollyRequest extends HTTPBase {
     );
 
     this.didRespond = false;
+    this.aborted = false;
     this.url = request.url;
     this.method = request.method.toUpperCase();
     this.body = request.body;
@@ -155,7 +156,7 @@ export default class PollyRequest extends HTTPBase {
     return this;
   }
 
-  async setup() {
+  async init() {
     // Trigger the `request` event
     await this._emit('request');
 
@@ -177,6 +178,10 @@ export default class PollyRequest extends HTTPBase {
       'Cannot respond to a request that already has a response.',
       !this.didRespond
     );
+
+    if (this.aborted) {
+      return;
+    }
 
     // Timestamp the response
     this.response.timestamp = timestamp();
@@ -206,6 +211,10 @@ export default class PollyRequest extends HTTPBase {
 
     // Trigger the `response` event
     await this._emit('response', this.response);
+  }
+
+  abort() {
+    this.aborted = true;
   }
 
   _overrideRecordingName(recordingName) {

--- a/packages/@pollyjs/core/src/server/handler.js
+++ b/packages/@pollyjs/core/src/server/handler.js
@@ -19,6 +19,7 @@ export default class Handler extends Map {
     this._eventEmitter = new EventEmitter({
       eventNames: [
         'error',
+        'abort',
         'request',
         'beforeReplay',
         'beforePersist',

